### PR TITLE
modify behaviour of the construct method

### LINKED ISSUE #897: Disable all validations
# Question

Please complete:
* OS: Ubuntu 18.04
* Python version `import sys; print(sys.version)`:
3.7.3 (default, Apr  3 2019, 19:16:38) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
* Pydantic version `import pydantic; print(pydantic.VERSION)`: 0.32.2

Hello anyone,
It possible to disable all validations of incoming data?

```py
import pydantic

class MyModel(pydantic.BaseModel):
    class Config:
        validation = False

...
```

### PR DESCRIPTION
**Pull request by samuelcolvin**
## Change Summary

Change `construct`, fix #897

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by dmontagu** [comment]
This is very close to what I want. A few points:

1) I think there might be a minor opportunity for performance improvement by caching the defaults map, but even without that this would obviously be a huge improvement.

2) It's still not an in-place swap for `BaseModel.__init__` due to the lack of a `**kwargs` argument.

3) Because validation does not occur on initialization, I think it is **especially** important to support a type-checkable API for this function, at least if we wanted to encourage it's use for known-valid inputs. I probably wouldn't feel comfortable using this function in my code without it.

    My mypy plugin can be used to typecheck inputs to `Model.__init__`, and it would be straightforward to adapt that same approach to any other method where the field values were passed as keyword arguments. But I think it would be hard/impossible to get type-checking to work nicely with a dict as the input argument like this.
    
    If I can't convince you to add a method with essentially the same signature as `BaseModel.__init__` that basically just calls the `construct` function with `fields_set=None`, that will be okay -- I'll just implement the method in a base class and write a custom extension to the existing (and hopefully eventually merged) mypy plugin. But I think many, many use cases could benefit from this API, and I would be harder-pressed to recommend the use of `construct` (as it is now) for this purpose now given how well supported type-checking is when using the `BaseModel.__init__` with PyCharm and mypy (soon).

**Comment by dmontagu** [comment]
To be more explicit, a method like this would be a lot easier to make type-safe:

```python
    @classmethod
    def safe_construct(cls: Type[Model], **data: Any) -> Model:
        # TODO: Cache defaults
        defaults = {name: deepcopy(field.default) for name, field in cls.__fields__.items() if not field.required}
        values = {**defaults, **data}
        fields_set = set(data.keys())
        return cls.construct(values, fields_set)
```

**Comment by samuelcolvin** [comment]
@dmontagu I'm fine with that, the only reason I didn't do that was that it avoids the possibility to provide `fields_set`, this isn't generally necessary, but there might be some situations where you want it.

E.g. you do something like:

```py
model_data = dict(model1)
...
# lots of process, RPC etc.
...
model2 = Model.construct(model_data)
```
But obviously `model_data` contains some default fields, so `model1.__fields_set__ != model2.__fields_set__`.

So I was allowing for

```py
model_data = dict(model1)
model_fields_set = model1.__fields_set__
...
# lots of process, RPC etc.
...
model2 = Model.construct(model_data, model_fields_set)
```
I know this won't be used much, but in theory it should be possible.

However I think we can have the best of both worlds:

since we have to trust the input data to `construct()` we know it can't have any keys starting with `_`, so we can have an extra keyword-only field `_fields_set` which can never conflict with field keys.

So the signature of construct becomes something like:

```py
def construct(cls: Type[Model], *, _fields_set: Optional['SetStr'] = None, **model_data: Any) -> Model:
```

Would this work for you?

Regarding caching `defaults` I don't think it's necessary, but if you prefer we can.

**Comment by dmontagu** [comment]
Starting with the `_` is a nice trick; that would work great for me. Let me see how much of a difference caching the defaults makes while profiling using my example above. It’s probably easiest to not cache and probably fast enough.

**Comment by dmontagu** [comment]
I did the profiling -- I get about a 30x speedup with this implementation over the with-validation-approach in my benchmarking script. I get about a 36x speedup if I cache the default values. But not caching removes a nuisance around mutable default values: we'd have to worry about deepcopying the cached values (I wasn't doing that when getting the 36x).

Given the "slow" version still offers a 30x speed improvement, and this method seems somewhat rarely used now, I'm happy with this (more maintainable) version. If this method starts getting heavier use, and if further performance improvements to it become a priority, we could always revisit.

File: pydantic/main.py [hunk 1]
```diff
@@ -414,14 +414,17 @@ def from_orm(cls: Type['Model'], obj: Any) -> 'Model':
         return m
 
     @classmethod
-    def construct(cls: Type['Model'], values: 'DictAny', fields_set: 'SetStr') -> 'Model':
+    def construct(cls: Type['Model'], _fields_set: Optional['SetStr'] = None, **values: Any) -> 'Model':
```
Comment (dmontagu) [inline pydantic/main.py]: @samuelcolvin Would it make sense to rename `cls` to `__pydantic_cls__` (similar to in `BaseModel.__init__`), or at least something that starts with a `_` to prevent keyword argument conflicts?

File: pydantic/main.py [hunk 2]
```diff
@@ -414,14 +414,17 @@ def from_orm(cls: Type['Model'], obj: Any) -> 'Model':
...
-        Creates a new model and set __dict__ without any validation, thus values should already be trusted.
-        Chances are you don't want to use this method directly.
+        Creates a new model setting __dict__ and __fields_set__ from trusted or pre-validated data.
+        Default values are respected, but no other validation is performed.
         """
+        defaults = {name: deepcopy(field.default) for name, field in cls.__fields__.items() if not field.required}
         m = cls.__new__(cls)
-        object.__setattr__(m, '__dict__', values)
-        object.__setattr__(m, '__fields_set__', fields_set)
+        object.__setattr__(m, '__dict__', {**defaults, **values})
```
Comment (dmontagu) [inline pydantic/main.py]: We can replace `defaults` with `cls.__defaults__` once #915 is merged; that can be a separate PR though if this is ready first.

**Comment by samuelcolvin** [comment]
I think this is ready, I'll merge it after #915 and the defaults tweak above.
